src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -0,0 +1,132 @@
+    // all outputs are unconditionally added too. However when an input is modified, the outputs may still
+    // be the same (for instance something changed elsewhere in a file that had no bearing on the produced
+    // output). In this case, the state table checks the results against the previously produced values,
+    // and any that are found to be the same instead get a cached state, meaning no new downstream work 
+    // will be produced for them. Thus a modified input is the only slot that can have differing output 
+    // states.
+
+    internal enum EntryState { Added, Removed, Modified, Cached };
+
+    internal interface IStateTable
+    {
+        IStateTable Compact();
+    }
+
+    /// <summary>
+    /// A data structure that tracks the inputs and output of an execution node
+    /// </summary>
+    /// <typeparam name="T">The type of the items tracked by this table</typeparam>
+    internal sealed class NodeStateTable<T> : IStateTable
+    {
+        internal static NodeStateTable<T> Empty { get; } = new NodeStateTable<T>(ImmutableArray<ImmutableArray<(T, EntryState)>>.Empty, isCompacted: true, exception: null);
+
+        //PROTOTYPE: there is no need to store the state per item
+        //           we can instead store one state per input, with
+        //           an optional set for modified states.
+        private readonly ImmutableArray<ImmutableArray<(T item, EntryState state)>> _states;
+
+        private readonly bool _isCompacted;
+
+        private readonly Exception? _exception;
+
+        private NodeStateTable(ImmutableArray<ImmutableArray<(T, EntryState)>> states, bool isCompacted, Exception? exception)
+        {
+            _states = states;
+            _isCompacted = isCompacted;
+            _exception = exception;
+        }
+
+        public bool IsFaulted { get => _exception is not null; }
+
+        public IEnumerator<(T item, EntryState state)> GetEnumerator()
+        {
+            return _states.SelectMany(s => s).GetEnumerator();
+        }
+
+        public IStateTable Compact()
+        {
+            if (_isCompacted)
+                return this;
+
+            var compacted = ArrayBuilder<ImmutableArray<(T, EntryState)>>.GetInstance();
+            foreach (var entry in _states)
+            {
+                // we have to keep empty entries
+                // we only remove all entries at once, so only need to check the first item
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
+                {
+                    compacted.Add(entry.SelectAsArray(e => (e.item, EntryState.Cached)));
+                }
+            }
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        // PROTOTYPE: this will be called to allow exceptions to flow through the graph
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
```

Comment:
>U [](http://example.com/codeflow?start=75&length=1)

Should this be `T`? (When would `T` and `U` be distinct?)

Reply:
When a 'downstream' `table<T>` is trying to get its values from an 'upstream' `table<U>`. `table<U>` is faulted, so we essentially just say, make a `table<T>` with the same fault.